x-pack/plugins/monitoring/server/kibana_monitoring/bulk_uploader.js

```diff
@@ -117,18 +117,39 @@ export class BulkUploader {
     return sendBulkPayload(this._client, this._interval, payload);
   }
 
+  static deepMergeUploadData(uploadData, collectorSet) {
```

Comment:
Does this method rely on the collectors being in a certain order in the set? If they come in random order each time, would that blow things up?

Reply:
It _shouldn't_ matter as we shouldn't have conflicting keys, but it might be worth detecting and throwing an error in that scenario. Is that what you meant?

Reply:
Mainly I'd be concerned that one collector's formatter needs to rely that a separate formatter has already run in order for a data structure to exist in a certain way for it to place itself. That isn't something that can be guaranteed easily.

I think this will be a piece to look at very closely. In thinking about how to implement this change, we should look into the possibility of *not* using the same data model as we have been. This data gets sent to the Elasticsearch Monitoring bulk endpoint, and it is possible for the endpoint to use an ingest script to rearrange the data, which would allow us to get away with some "breaking" changes.